### PR TITLE
[BillingSpec] break out one test but consolidate the rest

### PR DESCRIPTION
**Pull request by MatthewBemis**
Consolidate all of billing spec into 1 test with a newly created billing project. Also retry creating the billing project up to 3 times.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: If you changed a URL that is used elsewhere (e.g. in an email), comment about where it is used and ensure the dependent code is updated.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] **Submitter**: If you're adding new automated UI tests, review the test plan with QA
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Product Owner** sign off
- [ ] **Submitter**: Verify all tests go green, including CI tests and automated UI tests.
- [ ] **Submitter**: Squash commits and merge to develop. If adding test code, merge application code and test code at the same time.
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Comment by MatthewBemis** [comment]
jenkins retest

**Comment by MatthewBemis** [comment]
jEnKiNs ReTeSt

**Comment by MatthewBemis** [comment]
jenkins retest

**Comment by dvoet** [comment]
jenkins retest

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala [hunk 1]
```diff
@@ -23,110 +23,115 @@ class BillingSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Cl
...
+    logger.info(s"Creating billing project: $billingProjectName")
+
+    billingPage.createBillingProject(billingProjectName, Config.Projects.billingAccount)
+    register cleanUp Rawls.admin.deleteBillingProject(billingProjectName, UserInfo(OAuth2BearerToken(user.makeAuthToken().value), WorkbenchUserId("0"), WorkbenchEmail(user.email), 3600))(UserPool.chooseAdmin.makeAuthToken())
+
+    val statusOption = billingPage.waitForCreateDone(billingProjectName)
+
+    statusOption match {
+      case None | Some("failure") if trials > 1 =>
+        logger.info(s"failure or timeout creating project $billingProjectName, retrying ${trials-1} more times")
```
Comment (gpolumbo-broad) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:126]: Why do timeouts occur and how frequently do they occur?  For tests, simply retrying might be acceptable behavior, but in production, while not a failure, it's probably not ideal.  Should we do something more here to try to keep track of how frequently this timeout occurs and bubble that up as a metric that gets tracked?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala [hunk 2]
```diff
@@ -23,110 +23,115 @@ class BillingSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Cl
...
+          withWebDriver { implicit driver =>
+            withSignIn(user) { listPage =>
+              //BEGIN: Test creating billing project in UI
+              val billingPage = new BillingManagementPage().open
+              val billingProjectName = createNewBillingProject(user, billingPage)
+              //END: Test creating billing project in UI
+
+              //BEGIN: Test creating workspace
+              val workspaceName = "BillingSpec_makeWorkspace_" + randomUuid
+              register cleanUp api.workspaces.delete(billingProjectName, workspaceName)
```
Comment (gpolumbo-broad) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:71]: All new to me, and I don't have the source for the `CleanUp` trait yet.  However, I'm going to assume that this trait knows how to respond to just about any failure anywhere in this test to ensure that any and all created assets are properly cleaned up after a successful run or any failure?  

Without knowing details, is there anything else we can or should be cleaning up out of this test?  Users?  Buckets?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala [hunk 3]
```diff
@@ -23,110 +23,115 @@ class BillingSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Cl
...
+         * Clearly a lot of functionality bleeds between these test cases. In order to save runtime during
+         * the tests, we're combining all five. The idea is that if any piece of basic functionality
+         * in these tests was broken (creating a workspace in a new project, creating a new project, etc), then
+         * most of these tests would fail anyway so grouping them together isn't much worse.
+         *
+         * NOTE: this single test case MUST create a brand new billing project each time it's run. If we use a
+         * project from GPAlloc, this entire test becomes worthless and our test coverage for billing is garbage.
+         *
+         * E comes last because it makes the project unusable.
+         */
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:50]: Excellent documentation! Thank you!

**Review by gpolumbo-broad — APPROVED** [review]

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala [hunk 4]
```diff
@@ -23,110 +23,115 @@ class BillingSpec extends FreeSpec with WebBrowserSpec with UserFixtures with Cl
...
+              newBillingAccount shouldBe None
+              //END: should be able to change the billing account associated with the project
             }
           }
         }
       }
     }
   }
+
+  private def createNewBillingProject(user: Credentials, billingPage: BillingManagementPage, trials: Int = 3): String = {
```
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:115]: This bothers me for 2 reasons:
1. We will not notice these log messages and will therefore not be aware of changes in reliability. Getting logs into Kibana could help with this, especially if we can set up an alert if it happens too frequently.
2. With 3 retries, we are saying that we accept as low as ~33% reliability. A secondary check that gives us an alert while not failing the test could help with this.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala (hunk 1 above)
Comment (breilly2) [inline automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:126]: Maybe `warn` rather than `info`.

**Comment by aweng98** [comment]
can you break this test to two smaller sized tests? it helps with reducing overall execution time.

**Review by breilly2 — COMMENTED** [review]
I will approve as long as we have some visibility/tracking of the retries. Doesn't have to be immediate as long as we have a plan but we _must ***not***_ to do this.

**Comment by breilly2** [comment]
@aweng98 I disagree. Splitting it will not speed it up. Each one would need to create a new billing project which (other than running a workflow) is the time-consuming operation. Even running in parallel, splitting would not improve speed; also, It would decrease reliability.

**Comment by dvoet** [comment]
I have another PR coming in rawls to log failures. My hope is that these will eventually go to kibana and we can monitor/diagnose. As it stands there is no way other than error to get any attention.

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala (hunk 1 above)
Comment (dvoet) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:126]: does it really matter?

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala (hunk 2 above)
Comment (dvoet) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:71]: look in the workbench-libs repo for the source.

cleaning up the workspace cleans up the bucket (and so does cleaning up the project which is below). Users are cleaned when the fiab is torn down (fiab-stop).

File: automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala (hunk 1 above)
Comment (breilly2) [reply automation/src/test/scala/org/broadinstitute/dsde/firecloud/test/billing/BillingSpec.scala:126]: Doubtful. I probably tend to use info when I should use debug or trace so this feels more important to me than those. I don't really care a lot. I should have said "nit".

**Comment by dvoet** [comment]
I understand everyone's concerns over retrying this but I still think it is required for stability of the tests. We need to rebuild this functionality at some point soon anyway and maybe make it more reliable. In production there are 1200 projects and 59 failures. 5% failure rate, not great but ok. I don't want 5% failure rate in out tests.

**Review by breilly2 — APPROVED** [review]
Gah! A word disappeared from my previous review. I meant to say, "we must not _forget_ to do this." As long as we have a plan (and we don't forget to execute it), then I reluctantly agree that we're at the point where we need to do this for reliability reasons.

**Comment by aweng98** [comment]
@breilly2  In terms of testing, is it valid to test creating new billing-projects concurrently? two test files may or may not run at same time. but when they do run at same time, even though it's just creating 2 billing-projects, I think that's a good testing. I think splitting worth consideration.
but that's outside scope of this PR.

**Comment by breilly2** [comment]
@aweng98 Sounds like you're asking about a concurrency test for billing project creation. We don't currently have one of those. This PR doesn't change that fact or attempt to change it. We would have to write a new case which, as you point out, is outside the scope of this PR... and I don't have particularly strong feelings about the need for such a test.

This is also also separate from test run speed/throughput which was the reason you originally suggested splitting the test and the point that, in this case, I was disagreeing with.

**Pull request merged**